File: api/src/main/java/com/gkh/syntheticmonitor/model/SMActionAPI.java
package com.gkh.syntheticmonitor.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import groovy.lang.Writable;
import groovy.text.SimpleTemplateEngine;
import groovy.text.Template;
import lombok.*;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.ResourceAccessException;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.HashMap;
import java.util.Map;

@Data
@AllArgsConstructor
@NoArgsConstructor
@SuperBuilder
@Slf4j
public class SMActionAPI extends AbstractSMAction {

	// Todo: SSL certificate support?

	public final static String API = "API";
	public final static String METHOD_GET = "GET";
	public final static String METHOD_POST = "POST";
	public final static String METHOD_PUT = "PUT";
	public final static String METHOD_DELETE = "DELETE";

	@Singular
	private Map<String,String> requestHeaders = new HashMap<>();
	private String requestMethod;
	private String requestUrl;
	private String requestBody;
	private transient String requestBodyExpanded;
	private String expectedStatus;

	@JsonIgnore
	private transient String expandedUrl;

	@JsonIgnore
	private transient Map<String,String> expandedRequestHeaders;

	@Override
	@SneakyThrows
	public void resolveVariables(SMExecutionContext context) {
		this.expandedUrl =  UriComponentsBuilder.fromHttpUrl(this.requestUrl)
				.buildAndExpand(context.getVars())
				.toUriString();

		log.info("Expanded URL {}", this.expandedUrl);

		SimpleTemplateEngine engine = new groovy.text.SimpleTemplateEngine();
		if (this.requestBody != null && this.requestBody.contains("$")) {
			// Fixme: put this to abstract class
			Template template = engine.createTemplate(this.requestBody);
			Writable textTemplate = template.make(context.getVars());
			this.requestBodyExpanded  = textTemplate.toString();
		}
		else {
			this.requestBodyExpanded = this.requestBody;
		}

		this.expandedRequestHeaders = new HashMap<String,String>(this.requestHeaders);
		this.expandedRequestHeaders.forEach((k, v)->{
			try {
				if (v.contains("$")) {
					Template template = engine.createTemplate(v);
					Writable textTemplate = template.make(context.getVars());
					String expandedValue  = textTemplate.toString();
					this.expandedRequestHeaders.put(k,expandedValue);
				}
			} catch(Exception e) {
				e.printStackTrace();
			}
		});

	}

	@Override
	public String getDetails() {
		return this.getRequestMethod() + " " + this.requestUrl;
	}


	@Override
	public void execute(SMExecutionContext context) throws Exception  {

		log.info("API Test Execute");

		RestTemplate restTemplate = new RestTemplate();
		HttpHeaders httpHeaders = new HttpHeaders();

		// fixme
		if (expandedRequestHeaders!= null) {
			expandedRequestHeaders.keySet().forEach(key -> httpHeaders.set(key, expandedRequestHeaders.get(key)));
		}

		String status, content;

		try {
			HttpEntity entity = new HttpEntity(this.requestBodyExpanded, httpHeaders);
			HttpMethod httpMethod = getHttpMethod(requestMethod);

			ResponseEntity<String> response = restTemplate.exchange(this.expandedUrl, httpMethod, entity, String.class);

			status = Integer.toString(response.getStatusCode().value());
			content = response.getBody();

			log.info("Status: {}", status);
			log.info("Content: {}", content);
		} catch (ResourceAccessException e ) {
			status = "TIMEOUT";
			content = e.getMessage();
		}
		context.setContent(content);
		context.setStatus(status);
	}

	private static HttpMethod getHttpMethod(String method) throws Exception {
		if (method.equals(METHOD_GET)) return HttpMethod.GET;
		if (method.equals(METHOD_POST)) return HttpMethod.POST;
		if (method.equals(METHOD_PUT)) return HttpMethod.PUT;
		if (method.equals(METHOD_DELETE)) return HttpMethod.DELETE;
		throw new Exception("API Method not supported:" + method);
	}

	@Override
	public String getType() {
		return API;
	}

}

File: api/src/main/kotlin/com/gkh/syntheticmonitor/model/SMActionAPI.kt
package com.gkh.syntheticmonitor.model

import com.fasterxml.jackson.annotation.JsonIgnore
import lombok.extern.slf4j.Slf4j
import org.apache.logging.log4j.LogManager
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.web.client.ResourceAccessException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder
import java.util.*

@Slf4j
class SMActionAPI : AbstractSMAction() {

    companion object {
        var logger = LogManager.getLogger()

        // Todo: SSL certificate support?
        const val type = "API"
        const val METHOD_GET = "GET"
        const val METHOD_POST = "POST"
        const val METHOD_PUT = "PUT"
        const val METHOD_DELETE = "DELETE"

        @Throws(Exception::class)
        private fun getHttpMethod(method: String?): HttpMethod {
            if (method == METHOD_GET) return HttpMethod.GET
            if (method == METHOD_POST) return HttpMethod.POST
            if (method == METHOD_PUT) return HttpMethod.PUT
            if (method == METHOD_DELETE) return HttpMethod.DELETE
            throw Exception("API Method not supported:$method")
        }

    }

    var requestHeaders: MutableMap<String, String> = HashMap()
    var requestMethod: String? = null
    var requestUrl: String? = null
    var requestBody: String? = null

    override var type: String? = null
        get() = "API"

    @Transient
    var requestBodyExpanded: String? = null

    @JsonIgnore
    @Transient
    var urlExpanded: String? = null

    @JsonIgnore
    @Transient
    var requestHeadersExpanded: MutableMap<String, String>? = null

    override val expectedStatus: String? = null

    override fun expandInputParameters(context: SMExecutionContext) {
        this.expandURL(context)
        this.expandRequestBody(context)
        this.expandRequestHeaders(context)
    }

    fun expandRequestHeaders(context: SMExecutionContext) {
        val expandedRequestHeaders = HashMap<String, String>(requestHeaders)
        expandedRequestHeaders.forEach { (key, value) ->
            try {
                if (value.contains("$"))
                    expandedRequestHeaders[key] = evalTemplate(context, value)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        this.requestHeadersExpanded = expandedRequestHeaders
    }

    private fun expandURL(context: SMExecutionContext) {
        urlExpanded = UriComponentsBuilder.fromHttpUrl(this.requestUrl!!)
                .buildAndExpand(context.vars)
                .toUriString()
        logger.info("Expanded URL $urlExpanded")
    }

    fun expandRequestBody(context: SMExecutionContext) {
        requestBodyExpanded = if (this.requestBody?.contains("$") == true) {
            evalTemplate(context, requestBody!!)
        } else {
            this.requestBody
        }
        logger.info("Expanded request body $requestBodyExpanded")
    }

    override val details: String?
        get() = this.requestMethod + " " + requestUrl

    @Throws(Exception::class)
    override fun execute(context: SMExecutionContext) {
        logger.info("API Test Execute")
        val restTemplate = RestTemplate()
        val httpHeaders = HttpHeaders()

        this.requestHeadersExpanded?.let {
            this.requestHeadersExpanded!!.keys.forEach {
                httpHeaders[it] = requestHeadersExpanded!![it]
            }
        }

        var status: String?
        var content: String?

        try {
            val entity: HttpEntity<*> = HttpEntity<Any?>(requestBodyExpanded, httpHeaders)
            val httpMethod = getHttpMethod(requestMethod)
            val response = restTemplate.exchange(urlExpanded!!, httpMethod, entity, String::class.java)

            status = Integer.toString(response.statusCode.value())
            content = response.body

            logger.info("Status: $status")
            logger.info("Content: $content")
        } catch (e: ResourceAccessException) {
            status = "TIMEOUT"
            content = e.message.toString()
        }
        context.content = content
        context.status = status
    }


}